LP-1773 Add option for table field to open assessment block.

qasimgulzar opened the pull request.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] add checkbox template support to the plugin.
[NEW] add some basic styling.
[NEW] styleable html template.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] add generate table button to plugin.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] add radio template.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] add new type field to schema.py.
[NEW] integrate table view with assessment.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] plugin integration with ora backend.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] pre filled inputs for submissions.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] checkbox type
[PARTIAL][NEW] styling.

Commit:
checkbox fixes

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[FIX] checkbox fixes.

Commit:
Merge branch 'qasim/LP-1773' of https://github.com/philanthropy-u/edx-ora2 into qasim/LP-1773

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[FIX] caption styling, restrict changes after submission.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[FIX] plugin breaks.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] preserve open assessment table state.

Commit:
[PARTIAL][LP-1773] Add option for table field to open assessment block.
[NEW] add new `html_content` field to open assessment.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[FIX] disable add table button if assignment is published.
[NEW] add separator between add table and add prompt button.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[NEW] compiled js and remove default param values.
[UDPATE] Makefile

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
REFORMATTING.

emraan004 commented on openassessment/templates/openassessmentblock/edit/oa_edit_table_prompt.html:
```diff
@@ -0,0 +1,52 @@
+    </div>
+    <textarea class="openassessment_prompt_html_content setting-input is--hidden" data-w-type="html-control"></textarea>
+</li>
+{% endspaceless %}
```
let's add EOF

emraan004 commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,342 @@
+"use strict"
+
```
let's make constants for all selectors

mabdullahz commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -16,9 +16,15 @@ <h6 class="openassessment_container_header_title openassessment_prompt_header_ti
                 <div class="wrapper-comp-settings">
-                    <textarea class="openassessment_prompt_description setting-input" maxlength="10000">{{ prompt_description }}</textarea>
+                    <textarea class="openassessment_prompt_description setting-input" maxlength="10000" data-w-type="{{w_type}}">{{ prompt_description }}</textarea>
+                    <textarea class="openassessment_prompt_html_content setting-input is--hidden" data-w-type="html-control">{{html_content}}</textarea>
```
Let's add a space in `{{ asd }}`

mabdullahz commented on openassessment/templates/openassessmentblock/edit/oa_edit_table_prompt.html:
```diff
@@ -0,0 +1,52 @@
+    <div class="openassessment_container_header openassessment_prompt_header view-outline">
+        <div class="openassessment_container_header_title_box openassessment_prompt_header_title_box">
+            <h6 class="openassessment_container_header_title openassessment_prompt_header_title">{% trans "Prompt" %}</h6>
+            <p class="openassessment_container_guide openassessment_prompt_guide">{% trans "You cannot delete a prompt after the assignment has been released." %}</p>
```
Let's break this line into multiple.

mabdullahz commented on openassessment/templates/openassessmentblock/edit/oa_edit_table_prompt.html:
```diff
@@ -0,0 +1,52 @@
+            <h6 class="openassessment_container_header_title openassessment_prompt_header_title">{% trans "Prompt" %}</h6>
+            <p class="openassessment_container_guide openassessment_prompt_guide">{% trans "You cannot delete a prompt after the assignment has been released." %}</p>
+        </div>
+        <div class="openassessment_container_remove_button openassessment_prompt_remove_button"><h2>{% trans "Remove" %}</h2></div>
```
Let's move `<h2>` to a separate line.

mabdullahz commented on openassessment/templates/openassessmentblock/oa_submission_answer.html:
```diff
@@ -7,14 +7,37 @@
+                            $(question_div_id).initORATableCheckbox('checkbox-input', true);
+                            $(question_div_id).oaTable();
+                        });
+                    </script>
```
Let's check this again for when is it called and if that is correct, please.

qasimgulzar replied:
Done

mabdullahz commented on openassessment/templates/openassessmentblock/response/oa_response.html:
```diff
@@ -44,21 +44,49 @@ <h4 class="step__title">
+
+                                        {% else %}
                                         {{ part.prompt.description|linebreaks }}
+                                        {% endif %}
```
Let's properly indent this section

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[REMOVE] unwanted code.
[CHANGE] populate preselected input type to table row and column.
[CHANGE] use magic `constants`.
[CHANGE] add EOF to `oa_edit_table_prompt.html`
[UPDATE] minified javascript files.

qasimgulzar commented:
@emraan004 scenario related to rendering pre-selected input type from the select box is handled. I believe now there is no need to change the current flow under the current ticket.
other comments which were made are also resolved please have a look once again and let me know if there is anything else we can improve.

zahid-arbisoft commented:
There is small UI issue, not sure if it is within the scope. If user click generate table button as shown in following image. All the data in table get reset. I think this can be improved; either we can show popup before resetting data or can find any other appropriate way

![Screenshot 2020-02-10 at 1 19 22 PM](https://user-images.githubusercontent.com/31304505/74132343-2d332f80-4c08-11ea-89ea-17e13f0489e2.png)

zahid-arbisoft commented:
@qasimgulzar If I remove prompt and add table (which also include prompt, I am unable to save ORA in studio and always get following error. I guess having at least one prompt is mandatory, but if i keep default prompt and add table, we will have 2 prompts. Is this behaviour as per requirement

@ahmed-faraz46, @emraan004  

```
OpenAssessment Save Error
Validation error: Each example response for learner training must be unique.
```

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -61,7 +61,9 @@ Kwargs:
     this.templateElement = kwargs.templateElement;
+    this.templateTablePromptElement = kwargs.templateTablePromptElement;
     this.addButtonElement = kwargs.addButtonElement;
+    this.addTableButtonElement = kwargs.addTableButtonElement;
```
Since these are our own customisation, and there does not seem to be any ordering in which these variables are declared, nor is there any dependency between them, I think it would be a good idea to move these statements to the bottom of all variable declarations, together.

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -151,7 +155,49 @@ OpenAssessment.Container.prototype = {
+        var handlerItem = container.createContainerItem(containerItem);
+        handlerItem.addEventListeners();
+        handlerItem.addHandler();
+    },
```
Since these are changes in the core file, consider moving these into a separate file, and import from there. This would result in a decreased effort when we upgrade ORA, and is a good coding approach all in all.

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -151,7 +155,49 @@ OpenAssessment.Container.prototype = {
+            .toggleClass(this.containerItemClass, true)
+            .appendTo($(this.containerElement));
 
+        var tableConfiguration={
```
Kindly add the missing space before `{`

zahid-arbisoft replied:
Lets add space before and after `=`

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -151,7 +155,49 @@ OpenAssessment.Container.prototype = {
+            .appendTo($(this.containerElement));
 
+        var tableConfiguration={
+            change:function (table) {
```
Kindly add the missing space after `change:`

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -151,7 +155,49 @@ OpenAssessment.Container.prototype = {
+            }
+        }
+
+        var table=$(tablePrompt).oraTableBuilder(tableConfiguration);
```
Kindly add the missing spaces around `=`

zahid-arbisoft replied:
Lets add space before and after `=`

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -151,7 +155,49 @@ OpenAssessment.Container.prototype = {
+                .click(function(eventData) {
+                    var containerItem = container.createContainerItem(eventData.target);
+                    container.remove(containerItem);
+                });
```
Shouldn't this block be at the same level as `containerItem.find`?

zahid-arbisoft commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -16,9 +16,15 @@ <h6 class="openassessment_container_header_title openassessment_prompt_header_ti
                 <div class="wrapper-comp-settings">
-                    <textarea class="openassessment_prompt_description setting-input" maxlength="10000">{{ prompt_description }}</textarea>
+                    <textarea class="openassessment_prompt_description setting-input" maxlength="10000" data-w-type="{{w_type}}">{{ prompt_description }}</textarea>
+                    <textarea class="openassessment_prompt_html_content setting-input is--hidden" data-w-type="html-control">{{html_content}}</textarea>
```
Lets make `data-w-type` more descriptive because ORA's latest version has also introduced `data-type="{{ prompts_type }}`. If we have both data types then it is confusing what `w` means

qasimgulzar replied:
w stands for the widget I will add notations in docstring.

zahid-arbisoft commented on openassessment/templates/openassessmentblock/edit/oa_edit_prompt.html:
```diff
@@ -16,9 +16,15 @@ <h6 class="openassessment_container_header_title openassessment_prompt_header_ti
+                    <textarea class="openassessment_prompt_html_content setting-input is--hidden" data-w-type="html-control">{{html_content}}</textarea>
                 </div>
             </li>
+            {% if w_type == 'html-control' %}
```
Lets use descriptive name for type here as will `w_type `

qasimgulzar replied:
I have mentioned in plugins docstring.

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container.js:
```diff
@@ -151,7 +155,49 @@ OpenAssessment.Container.prototype = {
+                    container.remove(containerItem);
+                });
+        } else {
+            containerItem.find('.' + this.removeButtonClass).addClass('is--disabled');
```
Since this line is the same as the one in the `if` block, let's consider pre-setting it in a variable? Let me know what you think

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container_item.js:
```diff
@@ -112,6 +114,20 @@ OpenAssessment.Prompt.prototype = {
+    getHtmlContent: function () {
+        var htmlContent = $('.openassessment_prompt_html_content', this.element).first().val();
+        return htmlContent;
+    },
```
This is a core file, so please refer to the comment I made earlier about this and let's move this to a separate file and import from there if possible.

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_container_item.js:
```diff
@@ -112,6 +114,20 @@ OpenAssessment.Prompt.prototype = {
+    /**
+     *  Get prompt type from current textarea.
+     *
+     * @returns {string}
```
Please fix the spacing.

mabdullahz commented on openassessment/xblock/static/js/src/studio/oa_edit_prompts.js:
```diff
@@ -17,7 +17,9 @@ OpenAssessment.EditPromptsView = function(element, notifier) {
             templateElement: $("#openassessment_prompt_template", this.element).get(0),
+            templateTablePromptElement: $("#openassessment_table_prompt_template", this.element).get(0),
             addButtonElement: $("#openassessment_prompts_add_prompt", this.element).get(0),
+            addTableButtonElement: $("#openassessment_prompts_add_table", this.element).get(0),
```
Let's consider moving these lines together to the bottom?

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+"use strict"
```
Can we move this file somewhere else rather than putting it with the code files?

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+    const _this = this
+    var change_cb = function (table) {
+        console.info("Please overrider change in configuration object");
+    }
```
Let's confirm that once the files have been processed by webpack, the `console.info` does not execute, so we are sure this won't make it to prod

zahid-arbisoft commented on openassessment/templates/openassessmentblock/edit/oa_edit_table_prompt.html:
```diff
@@ -0,0 +1,52 @@
+<li id="ora-table-builder" class="openassessment_prompt ora-table-builder" data-prompt="{{ prompt_uuid }}">
+    <script type="text/html"  data-w-name="input" data-w-type="checkbox">
+        <label class="checkbox-input">
+            <label></label>
```
Are label tags redundant ?

qasimgulzar replied:
yes

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+        const __this = this;
+        this.controls = {
+            div: document.createElement(DIV),
+            input: document.createElement('input'),
```
Consider making a constant for `'input'`

RanaWaqar7550 replied:
**Nice to do this.**
Make an array like 
`var controllers = ['div', 'input', 'table', etc...];`
and pass it to a function i.e. makeControllers.

```
function makeControllers(controllers){
var result = {};
controller.forEach(function(c) {
result[c] = document.createElement(c);
});
return result;
}
```

`this.controls = makeControllers(controllers);`

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+                    $(_current_element).find(CHECKBOX_SELECTOR).prop('checked', $(this).data('checked'));
+                }
+
+
```
Please consider removing one line break here

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+                })
+            })
+        });
+        this.inputTypes = [
```
Please add a line break before this block.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            class: 'overrider_global_col_input',
+        });
+
+
```
Please consider removing one line break from here.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+        * */
+        var h2 = $(document.createElement('H2')), h5 = $(document.createElement('H5'));
+        var rows = [
+            $(this.controls[DIV].cloneNode()).attr({class: ROW}),
```
Let's move this into a variable (let's say `x`) and consider using `[x, x, x, x, x]` to simplify this. I was unable to find a reliable method to do array comprehension here. Or better yet, let's make a function that takes in an object of all `attr`s and use that here as well as in `var columns`.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            $(this.controls[DIV].cloneNode()).attr({class: COLUMN}),
+            $(this.controls[DIV].cloneNode()).attr({class: COLUMN}),
+            $(this.controls[DIV].cloneNode()).attr({class: COLUMN}),
+        ]
```
This looks like a lot of manual work which is not really scalable. It might be necessary but I don't have that much context to make a call, so please consider a cleaner approach to do this? I am more than happy to have a discussion on this.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+
+        this.editor_form_rows = rows;
+
+
```
Please remove on line break from here

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            $(this.controls[DIV].cloneNode()).attr({class: COLUMN}),
+        ]
+
+
```
Please remove one line break from here

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+    plugin.prototype.renderTable = function () {
+        var _this = this, rows = $(this.editor_rows).val(),
+            cols = $(this.editor_cols).val(),
+            contains_horizontal_headers = this.contains_horizontal_headers.find(CHECKBOX_SELECTOR).is(':checked'),
```
Since `':checked'` is being used 5 times in this file, let's move it into a constant

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            var option = this.controls['option'].cloneNode();
+            $(option).attr({value: r}).html('Row ' + (r + 1));
+            row_options.push(option);
+        }
```
We could create a function that takes a string (Row, Column) and an integer (for loop counter) as attributes and return us a list. Let me know what you think

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+        this.populate_input_type_select(this.col_input_type_select);
+        this.populate_input_type_select(this.row_input_type_select);
+
+
```
Let's remove one line break from here.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+
+        var rows_custom_input_types_div = $(this.editor_form_rows[3]).find(DIV);
+
+        const CUSTOM_INPUT_ROW_FIRST_COLUMN = 0, CUSTOM_INPUT_ROW_SECOND_COLUMN = 1;
```
This file seems to be a mixture of both ES5 and ES6. We are using `var` as well as `const` but not `() => {}` (arrow functions). I think the ORA module overall sticks with ES5 (correct me if I am wrong), so let's please remove any inconsistencies, and make sure that all the code we are writing is strictly in ES5.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+        });
+
+        // show custom column input type rows.
+        const CUSTOM_INPUT_TYPE_HEADER_ROW_INDEX = 2, CUSTOM_INPUT_TYPE_CONTROL_ROW_INDEX = 3;
```
After discussion with @RanaWaqar7550, we both think that there should be a block at the start of the scope for defining constants. Please avoid initialising constants right before they needed to be used, since this approach does not result in very readable code.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            $(thead).append(header_tr);
+            $(table).append(thead);
+        }
+        for (var r = 0; r < rows; r++) {
```
Let's add a line break here.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            $(th).attr({contenteditable: contenteditable});
+        })
+    }
+    return new plugin();
```
Let's please add proper spacing between code blocks.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+
+$.fn.initORATableCheckbox = function (checkboxClass, disabled) {
+    const _this = $(this);
+    const CHECKBOX_SELECTOR = 'input[type="checkbox"]';
```
Please avoid using `const` since this is supposed to be ES5.

mabdullahz commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+        return table.prop("outerHTML");
+    }
+    return new plugin(table);
+}
```
Line break please.

RanaWaqar7550 commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+
+    function plugin() {
+        this._this = _this;
+        const __this = this;
```
I don't quite understand this.
If you wanna make a new object then please use `Object.assign({}, _this)`.

qasimgulzar replied:
@RanaWaqar7550 I want to reflect changes to the context of parent closure here that's why I cannot make the new object.

RanaWaqar7550 commented on openassessment/xblock/static/js/lib/table_builder.js:
```diff
@@ -0,0 +1,356 @@
+            __this.controls[template_meta['wType']] = $($.parseHTML($(template).text()));
+            $(__this.controls[template_meta['wType']]).bind('DOMNodeInsertedIntoDocument', function () {
+                var _current_element = $(this);
+                $(this).find('label').first().text($(this).data('label')).click(function () {
```
please use `_current_element` instead of `$(this)`.

RanaWaqar7550 commented on openassessment/xblock/static/js/src/studio/oa_container_item.js:
```diff
@@ -92,7 +92,9 @@ OpenAssessment.Prompt.prototype = {
     getFieldValues: function() {
         var fields = {
-            description: this.description()
+            description: this.description(),
```
please don't make a variable and use return statement instead of fields.

qasimgulzar replied:
Done

RanaWaqar7550 commented on openassessment/xblock/static/js/src/studio/oa_container_item.js:
```diff
@@ -112,6 +114,20 @@ OpenAssessment.Prompt.prototype = {
+     * @returns {string}
+     */
+    promptType: function() {
+        var promptType = $('.openassessment_prompt_description', this.element).data('w-type');
```
please use return instead of making an extra variable.

qasimgulzar replied:
done

RanaWaqar7550 reviewed (changes requested):
Awesome effort!

Just a li'l improvement.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[FIX] PR comments are resolved.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
formatting issue

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
missing line break.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
move initializtion blocks

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
spacing

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[REFACTORING] overrider oa_container.js in separate file.

mabdullahz reviewed (approved):
Looks good to me! Great work @qasimgulzar!

RanaWaqar7550 reviewed (approved):
Awesome!

LGTM!

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[REFACTORING] overrider oa_container_item.js in separate file.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[REFACTORING] overrider oa_container_item.js in separate file.
[NEW] minified files.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[REFACTORING] move table plugin file to philu/lib folder.

Commit:
[COMPLETE][LP-1773] Add option for table field to open assessment block.
[REFACTORING] formatting.

emraan004 reviewed (approved):
👍
